### ISSUE TITLE
Fails to install python-olm-dev, No such file or directory #include <olm/pk.h>

### ISSUE THREAD

**Issue by bradyt** [poljar/weechat-matrix#74]
When I try to pip install the requirements, it fails on `python-olm-dev`. The `weechat-matrix` README suggests to install `libolm-dev` from apt, but if that is intended to provide the headers, I guess it doesn't.

I provide a Dockerfile below to suggest how to reproduce the issue, as well the output included in a folded section.

To solve the issue for me, It seems we can add `git clone https://git.matrix.org/git/olm.git` and `make install`.

``` Dockerfile
FROM ubuntu

RUN apt-get update && apt-get install -y \
        git \
        libolm-dev \
        python-pip \
        && rm -rf /var/lib/apt/lists/*

RUN python -m pip install --upgrade pip

RUN git clone https://github.com/poljar/weechat-matrix.git

WORKDIR weechat-matrix

ENV PATH="/root/.local/bin:${PATH}"

# python -m pip install -r requirements.txt
# results in errors
```

<details>
<summary>error</summary>

```
root@6af3facdaee6:/weechat-matrix# python -m pip install -r requirements.txt 
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
Collecting pyOpenSSL (from -r requirements.txt (line 1))
  Downloading https://files.pythonhosted.org/packages/01/c8/ceb170d81bd3941cbeb9940fc6cc2ef2ca4288d0ca8929ea4db5905d904d/pyOpenSSL-19.0.0-py2.py3-none-any.whl (53kB)
    100% |################################| 61kB 459kB/s 
Collecting typing (from -r requirements.txt (line 2))
  Downloading https://files.pythonhosted.org/packages/cc/3e/29f92b7aeda5b078c86d14f550bf85cff809042e3429ace7af6193c3bc9f/typing-3.6.6-py2-none-any.whl
Collecting webcolors (from -r requirements.txt (line 3))
  Downloading https://files.pythonhosted.org/packages/1d/44/c4902683be73beba20afd299705e11f0a753a01cc7f9d6a070841848605b/webcolors-1.8.1-py2.py3-none-any.whl
Collecting future (from -r requirements.txt (line 4))
  Downloading https://files.pythonhosted.org/packages/90/52/e20466b85000a181e1e144fd8305caf2cf475e2f9674e797b222f8105f5f/future-0.17.1.tar.gz (829kB)
    100% |################################| 829kB 2.0MB/s 
Collecting atomicwrites (from -r requirements.txt (line 5))
  Downloading https://files.pythonhosted.org/packages/52/90/6155aa926f43f2b2a22b01be7241be3bfd1ceaf7d0b3267213e8127d41f4/atomicwrites-1.3.0-py2.py3-none-any.whl
Collecting attrs (from -r requirements.txt (line 6))
  Downloading https://files.pythonhosted.org/packages/23/96/d828354fa2dbdf216eaa7b7de0db692f12c234f7ef888cc14980ef40d1d2/attrs-19.1.0-py2.py3-none-any.whl
Collecting logbook (from -r requirements.txt (line 7))
  Downloading https://files.pythonhosted.org/packages/f6/83/20fc0270614919cb799f76e32cf143a54c58ce2fa45c19fd38ac2e4f9977/Logbook-1.4.3.tar.gz (85kB)
    100% |################################| 92kB 2.7MB/s 
Collecting pygments (from -r requirements.txt (line 8))
  Downloading https://files.pythonhosted.org/packages/13/e5/6d710c9cf96c31ac82657bcfb441df328b22df8564d58d0c4cd62612674c/Pygments-2.3.1-py2.py3-none-any.whl (849kB)
    100% |################################| 849kB 3.4MB/s 
Collecting matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0 from git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0 (from -r requirements.txt (line 9))
  Cloning https://github.com/poljar/matrix-nio.git (to revision master) to /tmp/pip-install-5UL9Sk/matrix-nio
Collecting cryptography>=2.3 (from pyOpenSSL->-r requirements.txt (line 1))
  Downloading https://files.pythonhosted.org/packages/c3/c1/cf8665c955c9393e9ff0872ba6cd3dc6f46ef915e94afcf6e0410508ca69/cryptography-2.6.1-cp27-cp27mu-manylinux1_x86_64.whl (2.3MB)
    100% |################################| 2.3MB 14.9MB/s 
Requirement already satisfied: six>=1.5.2 in /usr/lib/python2.7/dist-packages (from pyOpenSSL->-r requirements.txt (line 1)) (1.11.0)
Collecting h11 (from matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0->-r requirements.txt (line 9))
  Downloading https://files.pythonhosted.org/packages/f9/f3/8e4cf5fa1a3d8bda942a0b1cf92f87815494216fd439f82eb99073141ba0/h11-0.8.1-py2.py3-none-any.whl (55kB)
    100% |################################| 61kB 3.9MB/s 
Collecting h2 (from matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0->-r requirements.txt (line 9))
  Downloading https://files.pythonhosted.org/packages/a6/b2/0348a08cce9980b15ef8607adc7f0534193fe25b9269daa0c327dc74f026/h2-3.1.0-py2.py3-none-any.whl (64kB)
    100% |################################| 71kB 4.3MB/s 
Collecting jsonschema (from matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0->-r requirements.txt (line 9))
  Downloading https://files.pythonhosted.org/packages/aa/69/df679dfbdd051568b53c38ec8152a3ab6bc533434fc7ed11ab034bf5e82f/jsonschema-3.0.1-py2.py3-none-any.whl (54kB)
    100% |################################| 61kB 7.0MB/s 
Collecting peewee (from matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0->-r requirements.txt (line 9))
  Downloading https://files.pythonhosted.org/packages/c9/e4/f82bd273a7c88e7a2f71652324039c7e382f2f88eced01b9799d6c936ad2/peewee-3.9.3.tar.gz (2.2MB)
    100% |################################| 2.2MB 2.1MB/s 
Collecting pycryptodome (from matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0->-r requirements.txt (line 9))
  Downloading https://files.pythonhosted.org/packages/41/80/216cd888906de20916c384d4345dd12fa4c475acd6b25af3f38449f727d5/pycryptodome-3.8.1-cp27-cp27mu-manylinux1_x86_64.whl (9.7MB)
    100% |################################| 9.7MB 427kB/s 
Collecting python-olm-dev (from matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0->-r requirements.txt (line 9))
  Downloading https://files.pythonhosted.org/packages/c1/37/02af98ce04dcfe473e5714703b4d5b1e4e77806a1c412c6c346310e132ac/python-olm-dev-3.0.tar.gz (238kB)
    100% |################################| 245kB 3.6MB/s 
Collecting unpaddedbase64 (from matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0->-r requirements.txt (line 9))
  Downloading https://files.pythonhosted.org/packages/d9/ee/6eff0c493c4cab61f17dfa65c058d1e8d2bf4ce33f44e1bc5c4834479dbc/unpaddedbase64-1.1.0-py2-none-any.whl
Requirement already satisfied: asn1crypto>=0.21.0 in /usr/lib/python2.7/dist-packages (from cryptography>=2.3->pyOpenSSL->-r requirements.txt (line 1)) (0.24.0)
Requirement already satisfied: enum34; python_version < "3" in /usr/lib/python2.7/dist-packages (from cryptography>=2.3->pyOpenSSL->-r requirements.txt (line 1)) (1.1.6)
Requirement already satisfied: ipaddress; python_version < "3" in /usr/lib/python2.7/dist-packages (from cryptography>=2.3->pyOpenSSL->-r requirements.txt (line 1)) (1.0.17)
Collecting cffi!=1.11.3,>=1.8 (from cryptography>=2.3->pyOpenSSL->-r requirements.txt (line 1))
  Downloading https://files.pythonhosted.org/packages/9d/6f/aea9f5559fb593da07ff34e67513bd62483b45715b4a5f5fae6a0a5792ea/cffi-1.12.2-cp27-cp27mu-manylinux1_x86_64.whl (413kB)
    100% |################################| 419kB 8.3MB/s 
Collecting hyperframe<6,>=5.2.0 (from h2->matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0->-r requirements.txt (line 9))
  Downloading https://files.pythonhosted.org/packages/19/0c/bf88182bcb5dce3094e2f3e4fe20db28a9928cb7bd5b08024030e4b140db/hyperframe-5.2.0-py2.py3-none-any.whl
Collecting hpack<4,>=2.3 (from h2->matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0->-r requirements.txt (line 9))
  Downloading https://files.pythonhosted.org/packages/8a/cc/e53517f4a1e13f74776ca93271caef378dadec14d71c61c949d759d3db69/hpack-3.0.0-py2.py3-none-any.whl
Collecting pyrsistent>=0.14.0 (from jsonschema->matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0->-r requirements.txt (line 9))
  Downloading https://files.pythonhosted.org/packages/8c/46/4e93ab8a379d7efe93f20a0fb8a27bdfe88942cc954ab0210c3164e783e0/pyrsistent-0.14.11.tar.gz (104kB)
    100% |################################| 112kB 2.2MB/s 
Collecting functools32; python_version < "3" (from jsonschema->matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0->-r requirements.txt (line 9))
  Downloading https://files.pythonhosted.org/packages/c5/60/6ac26ad05857c601308d8fb9e87fa36d0ebf889423f47c3502ef034365db/functools32-3.2.3-2.tar.gz
Requirement already satisfied: setuptools in /usr/lib/python2.7/dist-packages (from jsonschema->matrix-nio@ git+https://github.com/poljar/matrix-nio.git@master#egg=matrix-nio-0->-r requirements.txt (line 9)) (39.0.1)
Collecting pycparser (from cffi!=1.11.3,>=1.8->cryptography>=2.3->pyOpenSSL->-r requirements.txt (line 1))
  Downloading https://files.pythonhosted.org/packages/68/9e/49196946aee219aead1290e00d1e7fdeab8567783e83e1b9ab5585e6206a/pycparser-2.19.tar.gz (158kB)
    100% |################################| 163kB 8.4MB/s 
Building wheels for collected packages: future, logbook, matrix-nio, peewee, python-olm-dev, pyrsistent, functools32, pycparser
  Building wheel for future (setup.py) ... done
  Stored in directory: /root/.cache/pip/wheels/0c/61/d2/d6b7317325828fbb39ee6ad559dbe4664d0896da4721bf379e
  Building wheel for logbook (setup.py) ... done
  Stored in directory: /root/.cache/pip/wheels/a2/9f/6f/8c7a4ed6b9f6f3c98b742dbb0fd41fff3c130119c507376301
  Building wheel for matrix-nio (setup.py) ... done
  Stored in directory: /tmp/pip-ephem-wheel-cache-W69Qd_/wheels/05/35/77/9f7b8b779bc2cddf53647d8d68ef5f8c7731dbc30973e4df92
  Building wheel for peewee (setup.py) ... done
  Stored in directory: /root/.cache/pip/wheels/05/a3/fd/4753289244f7b40745f016555a95ba4529f11aecadf64c77f0
  Building wheel for python-olm-dev (setup.py) ... error
  Complete output from command /usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-5UL9Sk/python-olm-dev/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/pip-wheel-5zRBrC --python-tag cp27:
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-2.7
  creating build/lib.linux-x86_64-2.7/olm
  copying olm/utility.py -> build/lib.linux-x86_64-2.7/olm
  copying olm/group_session.py -> build/lib.linux-x86_64-2.7/olm
  copying olm/__version__.py -> build/lib.linux-x86_64-2.7/olm
  copying olm/pk.py -> build/lib.linux-x86_64-2.7/olm
  copying olm/account.py -> build/lib.linux-x86_64-2.7/olm
  copying olm/_compat.py -> build/lib.linux-x86_64-2.7/olm
  copying olm/session.py -> build/lib.linux-x86_64-2.7/olm
  copying olm/_finalize.py -> build/lib.linux-x86_64-2.7/olm
  copying olm/__init__.py -> build/lib.linux-x86_64-2.7/olm
  running build_ext
  generating cffi module 'build/temp.linux-x86_64-2.7/_libolm.c'
  creating build/temp.linux-x86_64-2.7
  building '_libolm' extension
  creating build/temp.linux-x86_64-2.7/build
  creating build/temp.linux-x86_64-2.7/build/temp.linux-x86_64-2.7
  x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fdebug-prefix-map=/build/python2.7-3hk45v/python2.7-2.7.15~rc1=. -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I/usr/include/python2.7 -c build/temp.linux-x86_64-2.7/_libolm.c -o build/temp.linux-x86_64-2.7/build/temp.linux-x86_64-2.7/_libolm.o
  build/temp.linux-x86_64-2.7/_libolm.c:496:18: fatal error: olm/pk.h: No such file or directory
           #include <olm/pk.h>
                    ^~~~~~~~~~
  compilation terminated.
  error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
  
  ----------------------------------------
  Failed building wheel for python-olm-dev
  Running setup.py clean for python-olm-dev
  Building wheel for pyrsistent (setup.py) ... done
  Stored in directory: /root/.cache/pip/wheels/83/59/9a/a037b9b3c3e93d9275ea0aff9d6064400f372879dfdab01afe
  Building wheel for functools32 (setup.py) ... done
  Stored in directory: /root/.cache/pip/wheels/b5/18/32/77a1030457155606ba5e3ec3a8a57132b1a04b1c4f765177b2
  Building wheel for pycparser (setup.py) ... done
  Stored in directory: /root/.cache/pip/wheels/f2/9a/90/de94f8556265ddc9d9c8b271b0f63e57b26fb1d67a45564511
Successfully built future logbook matrix-nio peewee pyrsistent functools32 pycparser
Failed to build python-olm-dev
Installing collected packages: pycparser, cffi, cryptography, pyOpenSSL, typing, webcolors, future, atomicwrites, attrs, logbook, pygments, h11, hyperframe, hpack, h2, pyrsistent, functools32, jsonschema, peewee, pycryptodome, python-olm-dev, unpaddedbase64, matrix-nio
  Found existing installation: cryptography 2.1.4
    Uninstalling cryptography-2.1.4:
      Successfully uninstalled cryptography-2.1.4
  Running setup.py install for python-olm-dev ... error
    Complete output from command /usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-5UL9Sk/python-olm-dev/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-yP8qBd/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-x86_64-2.7
    creating build/lib.linux-x86_64-2.7/olm
    copying olm/utility.py -> build/lib.linux-x86_64-2.7/olm
    copying olm/group_session.py -> build/lib.linux-x86_64-2.7/olm
    copying olm/__version__.py -> build/lib.linux-x86_64-2.7/olm
    copying olm/pk.py -> build/lib.linux-x86_64-2.7/olm
    copying olm/account.py -> build/lib.linux-x86_64-2.7/olm
    copying olm/_compat.py -> build/lib.linux-x86_64-2.7/olm
    copying olm/session.py -> build/lib.linux-x86_64-2.7/olm
    copying olm/_finalize.py -> build/lib.linux-x86_64-2.7/olm
    copying olm/__init__.py -> build/lib.linux-x86_64-2.7/olm
    running build_ext
    generating cffi module 'build/temp.linux-x86_64-2.7/_libolm.c'
    creating build/temp.linux-x86_64-2.7
    building '_libolm' extension
    creating build/temp.linux-x86_64-2.7/build
    creating build/temp.linux-x86_64-2.7/build/temp.linux-x86_64-2.7
    x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fdebug-prefix-map=/build/python2.7-3hk45v/python2.7-2.7.15~rc1=. -fstack-protector-strong -Wformat -Werror=format-security -fPIC -I/usr/include/python2.7 -c build/temp.linux-x86_64-2.7/_libolm.c -o build/temp.linux-x86_64-2.7/build/temp.linux-x86_64-2.7/_libolm.o
    build/temp.linux-x86_64-2.7/_libolm.c:496:18: fatal error: olm/pk.h: No such file or directory
             #include <olm/pk.h>
                      ^~~~~~~~~~
    compilation terminated.
    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
    
    ----------------------------------------
Command "/usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-5UL9Sk/python-olm-dev/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-yP8qBd/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-install-5UL9Sk/python-olm-dev/
```

</details>

**Comment by poljar** [comment]
I presume this is fixed as well since https://github.com/poljar/python-olm/issues/8 got resolved?

**Comment by bradyt** [comment]
Is the [`libolm-dev` package](https://packages.ubuntu.com/search?keywords=libolm-dev) too old even on Ubuntu 18.10, being version 2.2.2, rather than 3?

My impression is that current issue topic was caused by my not having installed olm, whereas the linked issue topic was caused by the old `libolm-dev` existing on my system.

Assuming all of that is correct, perhaps something like the following would make things smoother for users such as me?

``` diff
modified   README.md
@@ -19,12 +19,7 @@ support is still experimental.
 
 1. Install libolm
 
-    - Ubuntu
-        ```
-        sudo apt-get install libolm-dev
-        ```
-
-    - Debian (see https://git.matrix.org/git/olm)
+    - Debian/Ubuntu (see https://git.matrix.org/git/olm)
 
     - Archlinux based distribution (see https://aur.archlinux.org/packages/libolm/)
       use your favorite pacman frontend with AUR support (yaourt, yay, pikaur, …)
```

**Comment by poljar** [comment]
Yes, 3.0 is needed for the current python-olm version. This issue was caused by not having libolm 3.0 development headers installed in a place where the compiler can find them, the linked issue was caused by using the correct headers while compiling, yet linking to a 2.X version of libolm.

I don't know if libolm 3.0 is available in any repo for Debian/Ubuntu. If you think it's not I'll gladly merge a patch that changes the readme to reflect that.

**Comment by poljar** [comment]
Fixed by #75.

**Comment by zyphlar** [comment]
For future reference, in order to fix this in ubuntu 18.04 I had to manually download and install the DEB packages for libolm3 and libolm-dev v3:

https://pkgs.org/download/libolm

**Comment by jrhodnik** [comment]
I, too, had issues with this on Ubuntu 18.04. Had to manually download deb's from [here](https://packages.ubuntu.com/focal/libolm-dev) and [here](https://packages.ubuntu.com/focal/libolm3) and install them with `sudo dpkg -i [debname].deb`.

**Comment by xanoni** [comment]
Same problem in Debian Buster ... but Bullseye has libolm3

**Comment by F8ER** [comment]
> Fixed by #75.

Encountered the same issue. A solution: `sudo apt-get install -y libolm3 libolm-dev`.
